### PR TITLE
gas optimized loops

### PR DESCRIPTION
**Pull request by xtremetom**
Removed the need for two loops in `read()` and implemented `unchecked{}` where appropriate to help reduce gas on other loops. 
Further gas savings could be realized if we utilize do>while in place of for loops

### CODE REVIEW

File: packages/contracts/src/File.sol [hunk 1]
```diff
@@ -12,26 +12,29 @@ struct File {
 }
 
 function read(File memory file) view returns (string memory contents) {
-    address[] memory pointers = new address[](file.contents.length);
-    for (uint256 i = 0; i < file.contents.length; i++) {
-        pointers[i] = file.contents[i].pointer;
-    }
+
+    Content[] memory sContents = file.contents;
```
Comment (holic) [inline packages/contracts/src/File.sol:16]: what does `s` in `sContents` mean?

File: packages/contracts/src/File.sol [hunk 2]
```diff
@@ -12,26 +12,29 @@ struct File {
...
         // - get address
         // - get data size
         // - get code and add to contents
         // - update total size
-        let pointer := 0
+
         for { let i := 0 } lt(i, len) { i := add(i, 1) } {
-            pointer := mload(add(pointers, add(0x20, mul(i, 0x20))))
+            sContent := mload(add(sContents, add(0x20, mul(i, 0x20))))
+            pointer := mload(add(sContent, 0x20))
```
Comment (holic) [inline packages/contracts/src/File.sol:36]: Thanks for this! Was not sure how to access parts of the array-of-structs in assembly.

File: packages/contracts/src/File.sol (hunk 1 above)
Comment (xtremetom) [reply packages/contracts/src/File.sol:16]: Identifier for a struct

File: packages/contracts/src/File.sol (hunk 2 above)
Comment (xtremetom) [reply packages/contracts/src/File.sol:36]: :)

File: packages/contracts/src/File.sol [hunk 3]
```diff
@@ -12,26 +12,28 @@ struct File {
 }
 
 function read(File memory file) view returns (string memory contents) {
-    address[] memory pointers = new address[](file.contents.length);
-    for (uint256 i = 0; i < file.contents.length; i++) {
-        pointers[i] = file.contents[i].pointer;
-    }
+    Content[] memory chunks = file.contents;
```
Comment (holic) [inline packages/contracts/src/File.sol:15]: There must be a way to access this part of the `file` object in assembly without having to create a reference to it here.

**Pull request merged**